File: CS/PersonalOrganizer/Common/ViewModel/DocumentsViewModel.cs
using System;
using System.Linq;
using System.ComponentModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.DataAnnotations;
using PersonalOrganizer.Common.Utils;
using PersonalOrganizer.Common.DataModel;

namespace PersonalOrganizer.Common.ViewModel {
    public abstract class DocumentsViewModel<TModule, TUnitOfWork>
        where TModule : ModuleDescription<TModule>
        where TUnitOfWork : IUnitOfWork {

        protected readonly IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory;

        protected DocumentsViewModel(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory) {
            this.unitOfWorkFactory = unitOfWorkFactory;
            Modules = CreateModules().ToArray();
            foreach(var module in Modules)
                Messenger.Default.Register<NavigateMessage<TModule>>(this, module, x => Show(x.Token));
        }

        protected IDocumentManagerService DocumentManagerService { get { return this.GetService<IDocumentManagerService>(); } }

        protected IDocumentManagerService WorkspaceDocumentManagerService { get { return this.GetService<IDocumentManagerService>("WorkspaceDocumentManagerService"); } }

        public TModule[] Modules { get; private set; }

        protected virtual TModule DefaultModule { get { return Modules.First(); } }

        public virtual TModule SelectedModule { get; set; }

        public virtual TModule ActiveModule { get; protected set; }

        public void SaveAll() {
            Messenger.Default.Send(new SaveAllMessage());
        }

        public void OnClosing(CancelEventArgs cancelEventArgs) {
            Messenger.Default.Send(new CloseAllMessage(cancelEventArgs));
        }

        public void Show(TModule module) {
            if(module == null || DocumentManagerService == null)
                return;
            IDocument document = DocumentManagerService.FindDocumentByIdOrCreate(module, x => CreateDocument(module));
            document.Show();
        }

        protected bool IsLoaded { get; private set; }

        public virtual void OnLoaded() {
            IsLoaded = true;
            DocumentManagerService.ActiveDocumentChanged += OnActiveDocumentChanged;
            Show(DefaultModule);
        }

        void OnActiveDocumentChanged(object sender, ActiveDocumentChangedEventArgs e) {
            ActiveModule = e.NewDocument == null ? null : e.NewDocument.Id as TModule;
        }

        protected virtual void OnSelectedModuleChanged(TModule oldModule) {
            if(IsLoaded)
                Show(SelectedModule);
        }

        protected virtual void OnActiveModuleChanged(TModule oldModule) {
            SelectedModule = ActiveModule;
        }

        IDocument CreateDocument(TModule module) {
            var document = DocumentManagerService.CreateDocument(module.DocumentType, null, this);
            document.Title = GetModuleTitle(module);
            document.DestroyOnClose = false;
            return document;
        }

        protected virtual string GetModuleTitle(TModule module) {
            return module.ModuleTitle;
        }

        public void PinPeekCollectionView(TModule module) {
            if(WorkspaceDocumentManagerService == null)
                return;
            IDocument document = WorkspaceDocumentManagerService.FindDocumentByIdOrCreate(module, x => CreatePinnedPeekCollectionDocument(module));
            document.Show();
        }

        IDocument CreatePinnedPeekCollectionDocument(TModule module) {
            var document = WorkspaceDocumentManagerService.CreateDocument("PeekCollectionView", module.CreatePeekCollectionViewModel());
            document.Title = module.ModuleTitle;
            return document;
        }

        protected Func<TModule, object> GetPeekCollectionViewModelFactory<TEntity, TPrimaryKey>(Func<TUnitOfWork, IRepository<TEntity, TPrimaryKey>> getRepositoryFunc) where TEntity : class {
            return module => PeekCollectionViewModel<TModule, TEntity, TPrimaryKey, TUnitOfWork>.Create(module, unitOfWorkFactory, getRepositoryFunc).SetParentViewModel(this);
        }

        protected abstract TModule[] CreateModules();

        protected TUnitOfWork CreateUnitOfWork() {
            return unitOfWorkFactory.CreateUnitOfWork();
        }

        public virtual NavigationPaneVisibility NavigationPaneVisibility { get; set; }
    }

    public abstract partial class ModuleDescription<TModule> where TModule : ModuleDescription<TModule> {
        Func<TModule, object> peekCollectionViewModelFactory;

        object peekCollectionViewModel;

        protected ModuleDescription(string title, string documentType, string group, Func<TModule, object> peekCollectionViewModelFactory) {
            ModuleTitle = title;
            ModuleGroup = group;
            DocumentType = documentType;
            this.peekCollectionViewModelFactory = peekCollectionViewModelFactory;
        }

        public string ModuleTitle { get; private set; }

        public string ModuleGroup { get; private set; }

        public string DocumentType { get; private set; }

        public object PeekCollectionViewModel {
            get {
                if(peekCollectionViewModelFactory == null)
                    return null;
                if(peekCollectionViewModel == null)
                    peekCollectionViewModel = CreatePeekCollectionViewModel();
                return peekCollectionViewModel;
            }
        }

        public object CreatePeekCollectionViewModel() {
            return peekCollectionViewModelFactory((TModule)this);
        }
    }

    public enum NavigationPaneVisibility {
        Minimized,
        Normal,
        Off
    }
}

File: VB/PersonalOrganizer/Common/ViewModel/DocumentsViewModel.vb
Imports System
Imports System.Linq
Imports System.ComponentModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Mvvm.DataAnnotations
Imports PersonalOrganizer.Common.Utils
Imports PersonalOrganizer.Common.DataModel
Namespace PersonalOrganizer.Common.ViewModel
  Public MustInherit Class DocumentsViewModel(Of TModule As ModuleDescription(Of TModule), TUnitOfWork As IUnitOfWork)
    Private _IsLoaded As Boolean
    Private _Modules As TModule()
    Protected ReadOnly unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork)
    Protected Sub New(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork))
      Me.unitOfWorkFactory = unitOfWorkFactory
      _Modules = CreateModules().ToArray()
      For Each [module] In Modules
        Messenger.[Default].Register(Of NavigateMessage(Of TModule))(Me, [module], Sub(ByVal x) Show(x.Token))
      Next
    End Sub
    Protected ReadOnly Property DocumentManagerService As IDocumentManagerService
      Get
        Return Me.GetService(Of IDocumentManagerService)()
      End Get
    End Property
    Protected ReadOnly Property WorkspaceDocumentManagerService As IDocumentManagerService
      Get
        Return Me.GetService(Of IDocumentManagerService)("WorkspaceDocumentManagerService")
      End Get
    End Property
    Public ReadOnly Property Modules As TModule()
      Get
        Return _Modules
      End Get
    End Property
    Protected Overridable ReadOnly Property DefaultModule As TModule
      Get
        Return Modules.First()
      End Get
    End Property
    Public Overridable Property SelectedModule As TModule
    Public Overridable Property ActiveModule As TModule
    Public Sub SaveAll()
      Messenger.[Default].Send(New SaveAllMessage())
    End Sub
    Public Sub OnClosing(ByVal cancelEventArgs As CancelEventArgs)
      Messenger.[Default].Send(New CloseAllMessage(cancelEventArgs))
    End Sub
    Public Sub Show(ByVal [module] As TModule)
      If [module] Is Nothing OrElse DocumentManagerService Is Nothing Then
        Return
      End If
      Dim document As IDocument = DocumentManagerService.FindDocumentByIdOrCreate([module], Function(ByVal x) CreateDocument([module]))
      document.Show()
    End Sub
    Protected ReadOnly Property IsLoaded As Boolean
      Get
        Return _IsLoaded
      End Get
    End Property
    Public Overridable Sub OnLoaded()
      _IsLoaded = True
      AddHandler DocumentManagerService.ActiveDocumentChanged, AddressOf OnActiveDocumentChanged
      Show(DefaultModule)
    End Sub
    Private Sub OnActiveDocumentChanged(ByVal sender As Object, ByVal e As ActiveDocumentChangedEventArgs)
      ActiveModule = If(e.NewDocument Is Nothing, Nothing, TryCast(e.NewDocument.Id, TModule))
    End Sub
    Protected Overridable Sub OnSelectedModuleChanged(ByVal oldModule As TModule)
      If IsLoaded Then
        Show(SelectedModule)
      End If
    End Sub
    Protected Overridable Sub OnActiveModuleChanged(ByVal oldModule As TModule)
      SelectedModule = ActiveModule
    End Sub
    Private Function CreateDocument(ByVal [module] As TModule) As IDocument
      Dim document = DocumentManagerService.CreateDocument([module].DocumentType, Nothing, Me)
      document.Title = GetModuleTitle([module])
      document.DestroyOnClose = False
      Return document
    End Function
    Protected Overridable Function GetModuleTitle(ByVal [module] As TModule) As String
      Return [module].ModuleTitle
    End Function
    Public Sub PinPeekCollectionView(ByVal [module] As TModule)
      If WorkspaceDocumentManagerService Is Nothing Then
        Return
      End If
      Dim document As IDocument = WorkspaceDocumentManagerService.FindDocumentByIdOrCreate([module], Function(ByVal x) CreatePinnedPeekCollectionDocument([module]))
      document.Show()
    End Sub
    Private Function CreatePinnedPeekCollectionDocument(ByVal [module] As TModule) As IDocument
      Dim document = WorkspaceDocumentManagerService.CreateDocument("PeekCollectionView", [module].CreatePeekCollectionViewModel())
      document.Title = [module].ModuleTitle
      Return document
    End Function
    Protected Function GetPeekCollectionViewModelFactory(Of TEntity As Class, TPrimaryKey)(ByVal getRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TEntity, TPrimaryKey))) As Func(Of TModule, Object)
      Return Function(ByVal [module]) PeekCollectionViewModel(Of TModule, TEntity, TPrimaryKey, TUnitOfWork).Create([module], unitOfWorkFactory, getRepositoryFunc).SetParentViewModel(Me)
    End Function
    Protected MustOverride Function CreateModules() As TModule()
    Protected Function CreateUnitOfWork() As TUnitOfWork
      Return unitOfWorkFactory.CreateUnitOfWork()
    End Function
    Public Overridable Property NavigationPaneVisibility As NavigationPaneVisibility
  End Class
  Public Partial MustInherit Class ModuleDescription(Of TModule As ModuleDescription(Of TModule))
    Private _DocumentType As String
    Private _ModuleGroup As String
    Private _ModuleTitle As String
    Private _peekCollectionViewModelFactory As Func(Of TModule, Object)
    Private _peekCollectionViewModel As Object
    Protected Sub New(ByVal title As String, ByVal documentType As String, ByVal group As String, ByVal peekCollectionViewModelFactory As Func(Of TModule, Object))
      _ModuleTitle = title
      _ModuleGroup = group
      Me._DocumentType = documentType
      Me._peekCollectionViewModelFactory = peekCollectionViewModelFactory
    End Sub
    Public ReadOnly Property ModuleTitle As String
      Get
        Return _ModuleTitle
      End Get
    End Property
    Public ReadOnly Property ModuleGroup As String
      Get
        Return _ModuleGroup
      End Get
    End Property
    Public ReadOnly Property DocumentType As String
      Get
        Return _DocumentType
      End Get
    End Property
    Public ReadOnly Property PeekCollectionViewModel As Object
      Get
        If _peekCollectionViewModelFactory Is Nothing Then
          Return Nothing
        End If
        If _peekCollectionViewModel Is Nothing Then
          _peekCollectionViewModel = CreatePeekCollectionViewModel()
        End If
        Return _peekCollectionViewModel
      End Get
    End Property
    Public Function CreatePeekCollectionViewModel() As Object
      Return _peekCollectionViewModelFactory(CType(Me, TModule))
    End Function
  End Class
  Public Enum NavigationPaneVisibility
    Minimized
    Normal
    Off
  End Enum
End Namespace
